ecoli/analysis/multigeneration/ribosome_crowding.py

```diff
@@ -0,0 +1,248 @@
+
+    # Get field metadata
+    try:
+        field_names = field_metadata(
+            conn,
+            config_sql,
+            "listeners__ribosome_data__target_prob_translation_per_transcript",
+        )
+    except Exception as e:
+        print(f"[ERROR] Error getting field metadata: {e}")
+        return
+
+    # First pass: Find overcrowded monomer indices
+    overcrowded_query = f"""
+    WITH unnested AS (
+        SELECT 
+            unnest(listeners__ribosome_data__actual_prob_translation_per_transcript) as actual,
+            unnest(listeners__ribosome_data__target_prob_translation_per_transcript) as target,
+            generate_subscripts(listeners__ribosome_data__target_prob_translation_per_transcript, 1) as idx
+        FROM ({history_sql})
+    )
+    SELECT DISTINCT idx
+    FROM unnested
+    WHERE target > actual
+    ORDER BY idx
+    LIMIT {MAX_NUMBER_OF_MONOMERS_TO_PLOT}
+    """
+
+    overcrowded_indices = [
+        row[0] - 1 for row in conn.execute(overcrowded_query).fetchall()
+    ]  # Convert to 0-based
+
+    if not overcrowded_indices:
+        print("[INFO] No overcrowded monomers found.")
+        return
+
+    print(f"[INFO] Found {len(overcrowded_indices)} overcrowded monomers")
+
+    # Second pass: Get data for overcrowded monomers only
+    actual_expr = named_idx(
+        "listeners__ribosome_data__actual_prob_translation_per_transcript",
+        [f"actual_{i}" for i in range(len(overcrowded_indices))],
+        [overcrowded_indices],
+    )
+
+    target_expr = named_idx(
+        "listeners__ribosome_data__target_prob_translation_per_transcript",
+        [f"target_{i}" for i in range(len(overcrowded_indices))],
+        [overcrowded_indices],
+    )
+
+    data_query = f"SELECT {actual_expr}, {target_expr}, time FROM ({history_sql})"
+    df = conn.execute(data_query).fetchdf()
+
+    # ----------------------------------------- #
+    # Prepare plot data following original format
+    plot_data = []
+    n_overcrowded_monomers = len(overcrowded_indices)
+    n_overcrowded_monomers_to_plot = min(
+        n_overcrowded_monomers, MAX_NUMBER_OF_MONOMERS_TO_PLOT
+    )
+
+    for i, idx in enumerate(overcrowded_indices):
+        if i >= MAX_NUMBER_OF_MONOMERS_TO_PLOT:
+            break
+
+        if idx < len(field_names):
+            monomer_id = field_names[idx]
+            gene_id = monomer_to_gene.get(monomer_id, "Unknown")
+
+            # Add target probabilities
+            for _, row in df.iterrows():
+                plot_data.append(
+                    {
+                        "Time_min": float(row["time"]),
+                        "Gene_ID": str(gene_id),
+                        "Probability_Type": "target",
+                        "Translation_Probability": float(row[f"target_{i}"]),
+                        "Plot_Order": i,
+                    }
+                )
+
+            # Add actual probabilities
+            for _, row in df.iterrows():
```

Comment:
You can do this all under a single for loop. A better solution might be to set this up to work using Polars `unpivot` (replacement for `melt`). If you name your columns like `{probability_type}__{gene_id}` using `named_idx`, you can probably `unpivot` then split the variable name column into separate probability type and gene ID columns using [`split_exact`](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.split_exact.html). If you have time, can you benchmark these two options? You can just create a reasonably sized synthetic dataset with say 20 gene IDs and 3000*32 rows (~32 generations of data).